File: Baekjoon/Class3/16.11727 2×n 타일링 2.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8");

const DIVISOR = 10007;

const n = Number(input);

const tilingCases = new Array(n + 1).fill(0); // 타뷸레이션
if(n >= 1) tilingCases[1] = 1;
if(n >= 2) tilingCases[2] = 3;
for(let i = 3; i <= n; i++)
    tilingCases[i] = (tilingCases[i - 2] * 2 + tilingCases[i - 1]) % DIVISOR;

console.log(tilingCases[n]);

File: Baekjoon/Class3/16.11727 2×n 타일링 2.py
import sys
input = sys.stdin.readline

DIVISOR = 10007

n = int(input())

tiling_cases = [0] * (n + 1) #타뷸레이션
if n >= 1:
    tiling_cases[1] = 1
if n >= 2:
    tiling_cases[2] = 3
for i in range(3, n + 1):
    tiling_cases[i] = (tiling_cases[i - 2] * 2 + tiling_cases[i - 1]) % DIVISOR

print(tiling_cases[n])